benchmarks/distributed/rpc/ps/benchmark.py

```diff
@@ -0,0 +1,562 @@
+from torch.distributed.rpc import TensorPipeRpcBackendOptions
+import torch.multiprocessing as mp
+from torch.nn.parallel import DistributedDataParallel as DDP
+import torch.optim as optim
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.distributed as c10d
+import torch.distributed.rpc as rpc
+
+# --------------------------- constants -----------------------------------------
+
+GLOO = "gloo"
+NCCL = "nccl"
+SPARSE = "sparse"
+INDICES = "indices"
+VALUES = "values"
+SIZE = "size"
+GRADIENT = "gradient"
+HOOK_METRIC = "hook_metric"
+FORWARD_METRIC = "foward_metric"
+BACKWARD_METRIC = "backward_metric"
+BATCH_LEVEL_METRIC = "batch_level_metric"
+RANK = "rank"
+TRAINER_COUNT = "trainer_count"
+USE_CUDA_RPC = "use_cuda_rpc"
+BATCH_MODE = "batch_mode"
+
+
+# --------------------------- basic helpers -----------------------------------------
+
+
+def get_name(rank, world_size):
+    if rank < world_size - 2:
+        return "trainer{}".format(rank)
+    elif rank == world_size - 2:
+        return "gs"
+    else:
+        return "master"
+
+# --------------------------- metrics -----------------------------------------
+
+
+def record_event(rank, metric_type, key, name, metrics):
+    event = torch.cuda.Event(enable_timing=True)
+    if metric_type not in metrics:
+        metrics[metric_type] = {}
+    metrics = metrics[metric_type]
+    if key in metrics:
+        assert "end" not in metrics[key]
+        metrics[key]["end"] = event
+    else:
+        metrics[key] = {
+            "name": name,
+            "start": event
+        }
+    with torch.cuda.device(rank):
+        event.record()
+
+# --------------------------- Parameter Server -----------------------------------------
+
+# TODO: fix implementation to allow for multiple gradient servers
+
+class GradientServer:
+
+    #TODO: find out how to use cuda kernel for sparse tensor addition / implement if there is no kernel
+    #TODO: metrics for computations in average_gradient
+
+    def __init__(self, rank, trainer_count, backend, use_cuda_rpc, batch_mode):
+        self.lock = threading.Lock()
+
+        self.rank = rank
+        self.trainer_count = trainer_count
+        self.use_cuda_rpc = use_cuda_rpc
+        self.backend = backend
+        self.batch_mode = batch_mode
+
+        self.futures = {}
+        self.gradient = {}
+        self.batch_number = 0
+
+
+    @staticmethod
+    @rpc.functions.async_execution
+    def average_gradient(gs_rref, rank, bp_loc, received_batch_number, **kwargs):
+        sparse_gradient = False
+        if SPARSE in kwargs and kwargs[SPARSE]:
+            assert INDICES in kwargs and kwargs[INDICES] is not None
+            assert VALUES in kwargs and kwargs[VALUES] is not None
+            assert SIZE in kwargs and kwargs[SIZE] is not None
+            sparse_gradient = True
+            gradient = torch.sparse_coo_tensor(kwargs[INDICES], kwargs[VALUES], kwargs[SIZE])
+        else:
+            gradient = kwargs[GRADIENT]
```

Comment:
do we need similar assert on kwargs and GRADIENT as did in the `SPARSE` branch?

Reply:
I have removed most of the assert statements in the code. I left only asserts in the initialization of the configurations object. I was testing out coding with asserts before submitting a pull request for review.